tests/legacy/bootstrap.php

```diff
@@ -99,6 +110,35 @@ private function initialize_code_hacker() {
 		CodeHacker::enable();
 	}
 
+	/**
+	 * Re-initialize the dependency injection engine.
+	 *
+	 * The dependency injection engine has been already initialized as part of the Woo initialization, but we need
+	 * to replace the registered read-only container with a fully configurable one for testing.
+	 * To this end we hack a bit and use reflection to grab the underlying container that the read-only one stores
+	 * in a private property.
+	 *
+	 * Additionally, we replace the legacy/function proxies with mockable versions to easily replace anything
+	 * in tests as appropriate.
+	 *
+	 * @throws \Exception The Container class doesn't have a 'container' property.
+	 */
+	private function initialize_dependency_injection() {
```

Comment:
I'm just wondering, since we've already initialized the container, wouldn't we have already initialized any legacy classes with the "wrong" dependencies? Is there some way we can hack a mockable container in BEFORE WooCommerce is initialized? Technically if we have a class in the same namespace and load it, the autoloader won't pull in the new class, I wonder if we can abuse that?

Reply:
In this case I'd take the "let's not worry about a problem that doesn't exist yet" approach. Initialising the container is an internal process that we can change later without breaking anything if the need arises.

Still, it will be useful to make the mental exercise of imagining how that scenario would go, so let's imagine that we have a legacy class that depends on a new-style class that we need to mock. As I see it there are two possibilities:

1. The legacy class is not a singleton. Then no problemo, we'll just generate a new instance for testing, right after we have mocked whatever needs to be mocked.

2. The legacy class is a singleton. Then we can change the code of the original class to generate the singleton with a `wc_get_container()->get(LegacyProxy::class)->get_instance_of(Myself::class)` instead of a `new Myself()`, and from that point we are free to instruct the `MockableLegacyMocker` to return a new instance, and then GOTO 1.

So: theoretically that's not a problem. But if at some point it is, then it'll be safe to do any changes we need.

Reply:
In hindsight we can just instantiate a new version of the class under test and it will use the initialized container just fine. Another terrible comment 😄